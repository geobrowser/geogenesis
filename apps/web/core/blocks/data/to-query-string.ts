import { SystemIds } from '@graphprotocol/grc-20';

import { OmitStrict } from '~/core/types';

import { Filter } from './filters';

export function queryStringFromFilters(
  filters: OmitStrict<Filter, 'valueName' | 'columnName' | 'relationValueTypes'>[]
): string {
  if (filters.length === 0) return '';

  const filtersAsStrings = filters
    .map(filter => {
      // Assume we can only filter by one type at a time for now
      if (filter.columnId === SystemIds.TYPES_ATTRIBUTE && filter.valueType === 'RELATION') {
        return `versionTypes: { some: { type: { entityId: {equalTo: "${filter.value}" } } } }`;
      }

      // We treat Name and Space as special filters even though they are not always
      // columns on the type schema for a table. We allow users to be able to filter
      // by name and space.
      if (filter.columnId === SystemIds.NAME_ATTRIBUTE && filter.valueType === 'TEXT') {
        // For the name we can just search for the name based on the indexed GeoEntity name
        return `name: { startsWithInsensitive: "${filter.value}" }`;
      }

      if (filter.columnId === SystemIds.SPACE_FILTER && filter.valueType === 'RELATION') {
        return `versionSpaces: {
          some: {
            spaceId: { equalTo: "${filter.value}" }
          }
        }`;
      }

      if (filter.valueType === 'TEXT') {
        // value is just the textValue of the triple
        return `triples: { some: { attributeId: { equalTo: "${filter.columnId}" }, textValue: { equalToInsensitive: "${filter.value}"} } }`;
      }

      if (filter.valueType === 'RELATION') {
        return `relationsByFromVersionId: {
                some: {
                  typeOf: { id: { equalTo: "${filter.columnId}" } }
                  toEntity: { id: { equalTo: "${filter.value}" } }
                }
              }`;
      }

      // We don't support other value types yet
      return null;
    })
    .flatMap(f => (f ? [f] : []));

  if (filtersAsStrings.length === 1) {
    return `${filtersAsStrings[0]}`;
  }

  // Wrap each filter expression in curly brackets
  const multiFilterQuery = filtersAsStrings.map(f => `{ ${f} }`).join(', ');

  return `and: [${multiFilterQuery}]`;
}
