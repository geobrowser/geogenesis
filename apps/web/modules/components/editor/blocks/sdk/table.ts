import { SYSTEM_IDS } from '@geogenesis/ids';

import { Entity } from '~/modules/entity';
import { ID } from '~/modules/id';
import { Triple } from '~/modules/triple';
import { Column, EntityValue, Entity as IEntity, Triple as ITriple, TripleValueType } from '~/modules/types';

export function upsertName({
  blockEntity,
  name,
  api,
}: {
  blockEntity: IEntity | null;
  name: string;
  api: {
    create: (triple: ITriple) => void;
    update: (triple: ITriple, oldTriple: ITriple) => void;
  };
}) {
  if (!blockEntity) return;

  const nameTriple = Entity.nameTriple(blockEntity?.triples ?? []);

  if (!nameTriple)
    return api.create(
      Triple.withId({
        attributeId: SYSTEM_IDS.NAME,
        entityId: blockEntity.id,
        entityName: name,
        attributeName: 'Name',
        space: blockEntity.nameTripleSpace ?? '',
        value: { type: 'string', id: ID.createValueId(), value: name },
      })
    );

  api.update({ ...nameTriple, value: { ...nameTriple.value, type: 'string', value: name } }, nameTriple);
}

// @TODO: Some of the functionality in here could be re-used for all blocks
export function createBlock({
  blockEntity,
  spaceId,
  api,
  parentEntityId,
  parentName,
  rowTypeEntityId,
  rowTypeEntityName,
}: {
  blockEntity: IEntity | null;
  spaceId: string;
  parentEntityId: string;
  parentName: string;
  rowTypeEntityId: string;
  rowTypeEntityName: string;
  api: { create: (triple: ITriple) => void };
}) {
  // 1. Create parent triple
  const existingParentTypeTriple = blockEntity?.triples.find(t => t.attributeId === SYSTEM_IDS.PARENT_ENTITY);

  const entityId = blockEntity?.id ?? ID.createEntityId();
  const entityName = blockEntity?.name ?? `Table Block ${entityId}`;

  if (!existingParentTypeTriple) {
    api.create(
      Triple.withId({
        space: spaceId,
        entityId,
        entityName,
        attributeId: SYSTEM_IDS.PARENT_ENTITY,
        attributeName: 'Parent Entity',
        value: { id: parentEntityId, type: 'entity', name: parentName },
      })
    );
  }

  // 2. Create row type triple
  const existingRowTypeTriple = blockEntity?.triples.find(t => t.attributeId === SYSTEM_IDS.ROW_TYPE);

  if (!existingRowTypeTriple) {
    api.create(
      Triple.withId({
        space: spaceId,
        entityId: blockEntity?.id ?? ID.createEntityId(),
        entityName: blockEntity?.name ?? `Table Block ${blockEntity?.id}`,
        attributeId: SYSTEM_IDS.ROW_TYPE,
        attributeName: 'Row Type',
        value: { id: rowTypeEntityId, type: 'entity', name: rowTypeEntityName },
      })
    );
  }

  // 3. Create blocktype triple
  const blockTypeValue: EntityValue = { id: SYSTEM_IDS.TABLE_BLOCK, type: 'entity', name: 'Table Block' };
  const existingBlockTypeTriple = blockEntity?.triples.find(t => t.attributeId === SYSTEM_IDS.TYPES);

  if (!existingBlockTypeTriple) {
    api.create(
      Triple.withId({
        space: spaceId,
        entityId: entityId,
        entityName: entityName,
        attributeId: SYSTEM_IDS.TYPES,
        attributeName: 'Types',
        value: blockTypeValue,
      })
    );
  }

  // 4. Create name triple
  const existingNameTriple = blockEntity?.triples.find(t => t.attributeId === SYSTEM_IDS.TYPES);

  if (!existingNameTriple) {
    api.create(
      Triple.withId({
        space: spaceId,
        entityId: entityId,
        entityName: entityName,
        attributeId: SYSTEM_IDS.TYPES,
        attributeName: 'Types',
        value: { id: ID.createValueId(), type: 'string', value: entityName },
      })
    );
  }
}

/**
 * Takes the table filters and converts them to the GraphQL string used to
 * query the table using the filters. We include the typeId from the table
 * in the graphql string to make sure we're filtering by the correct type.
 *
 * e.g. these filters
 * ```ts
 * const filters = [{
 *   columnId: 'type',
 *   value: 'd73a9e43-923e-4102-87da-5d3176ac9df2', // entity ID for 'Blockchain'
 *   valueType: 'entity',
 *  },
 *  {
 *   columnId: 'type',
 *   value: '48a331d1-a6d6-49ca-bc23-78f3378eb959', // entity ID for 'Layer 1'
 *   valueType: 'entity',
 * }]
 * ```
 *
 * would output to
 * ```ts
 * `{
 *    and: [
 *      {entityOf_: {attribute: "type", entityValue: "d73a9e43-923e-4102-87da-5d3176ac9df2"}},
 *      {entityOf_: {attribute: "type", entityValue: "48a331d1-a6d6-49ca-bc23-78f3378eb959"}},
 *      name: "Bitcoin"
 *    ]
 * }`
 * ```
 */
export function createGraphQLStringFromFilters(
  filters: {
    columnId: string;
    valueType: TripleValueType;
    value: string;
  }[],
  typeId: string | null
): string {
  if (!typeId) return '';
  if (filters.length === 0) return `{typeIds_contains_nocase: ["${typeId}"]}`;

  const filtersAsStrings = filters
    .map(filter => {
      if (filter.columnId === SYSTEM_IDS.NAME && filter.valueType === 'string') {
        // For the name we can just search for the name based on the indexed GeoEntity name
        return `name_starts_with_nocase: "${filter.value}"`;
      }

      if (filter.valueType === 'entity') {
        // value is the ID of the relation
        return `entityOf_: {attribute: "${filter.columnId}", entityValue: "${filter.value}"}`;
      }

      if (filter.valueType === 'string') {
        // value is just the stringValue of the triple
        return `entityOf_: {attribute: "${filter.columnId}", stringValue_starts_with_nocase: "${filter.value}"}`;
      }

      // We don't support other value types yet
      return null;
    })
    .flatMap(f => (f ? [f] : []));

  if (filtersAsStrings.length === 1) {
    return `{typeIds_contains_nocase: ["${typeId}"], ${filtersAsStrings[0]}}`;
  }

  // Wrap each filter expression in curly brackets
  const multiFilterQuery = filtersAsStrings.map(f => `{${f}}`).join(', ');

  return `{and: [{typeIds_contains_nocase: ["${typeId}"]}, ${multiFilterQuery}]}`;
}

/**
 * Takes the graphQL string representing the TableBlock filters and maps them to the
 * application/UI representation of the filters.
 *
 * Turns this:
 * ```ts
 * {
 *    and: [
 *     {
 *       typeIds_contains_nocase: ["type-id"]
 *     },
 *     {
 *       entityOf_: {attribute: "type", stringValue_starts_with_nocase: "Value 1"}
 *     },
 *     {
 *       entityOf_: {attribute: "type", entityValue: "id 1"}
 *     },
 *     {
 *       name_starts_with_nocase: "id 1"
 *     }
 *   ]
 * }
 * ```
 *
 * into this:
 *```ts
 * [
 *  {
 *     columnId: 'type',
 *     columnName: 'Type',
 *     valueType: 'string',
 *     value: 'Value 1'
 *   },
 *   {
 *     columnId: 'type',
 *     columnName: 'Type',
 *     valueType: 'entity',
 *     value: 'id 1'
 *   },
 *   {
 *     columnId: 'name',
 *     columnName: 'Name',
 *     valueType: 'string',
 *     value: 'id 1'
 *   }
 * ]
 * ```
 */
export async function createFiltersFromGraphQLString(
  graphQLString: string,
  fetchEntity: (entityId: string) => Promise<IEntity | null>
): Promise<
  {
    columnId: string;
    valueType: TripleValueType;
    value: string;
    valueName: string | null;
  }[]
> {
  const filters: {
    columnId: string;
    valueType: TripleValueType;
    value: string;
    valueName: string | null;
  }[] = [];

  // Parse a name query from the filter
  const nameRegex = /name_starts_with_nocase\s*:\s*"([^"]*)"/;
  const nameMatch = graphQLString.match(nameRegex);
  const nameValue = nameMatch ? nameMatch[1] : null;

  if (nameValue) {
    filters.push({
      columnId: SYSTEM_IDS.NAME,
      valueType: 'string',
      value: nameValue,
      valueName: null,
    });
  }

  // Parse all entity relationship queries from the filter
  const entityValueRegex = /entityOf_\s*:\s*{\s*attribute\s*:\s*"([^"]*)"\s*,\s*entityValue\s*:\s*"([^"]*)"\s*}/g;

  for (const match of graphQLString.matchAll(entityValueRegex)) {
    const attribute = match[1];
    const entityValue = match[2];

    if (attribute && entityValue) {
      const maybeEntity = await fetchEntity(entityValue);

      if (maybeEntity) {
        filters.push({
          columnId: attribute,
          valueType: 'entity',
          value: entityValue,
          valueName: maybeEntity.name,
        });
      }
    }
  }

  // Parse all string queries from the filter
  const stringValueRegex =
    /entityOf_\s*:\s*{\s*attribute\s*:\s*"([^"]*)"\s*,\s*stringValue_starts_with_nocase\s*:\s*"([^"]*)"\s*}/g;

  for (const match of graphQLString.matchAll(stringValueRegex)) {
    const attribute = match[1];
    const stringValue = match[2];

    if (attribute && stringValue) {
      filters.push({
        columnId: attribute,
        valueType: 'string',
        value: stringValue,
        valueName: null,
      });
    }
  }

  return filters;
}
