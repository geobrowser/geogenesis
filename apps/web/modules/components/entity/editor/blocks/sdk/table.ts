export function createTableBlock() {
  return;
}

export function editTableBlock() {
  return;
}

export function deleteTableBlock() {
  return;
}
