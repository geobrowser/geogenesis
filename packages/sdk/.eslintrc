{
  "env": {
    "es2021": true,
    "node": true
  },
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/recommended",
    "prettier"
  ],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx"]
    }
  ],
  "ignorePatterns": ["node_modules/**/*", ".turbo/**/*", ".cache/**/*"],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": "latest",
    "sourceType": "module"
  },
  "plugins": ["react", "@typescript-eslint"],
  "rules": {
    "@typescript-eslint/no-non-null-assertion": 0,
    "@typescript-eslint/no-unused-vars": 1,
    "@typescript-eslint/no-empty-interface": 0,
    "no-undef": 1,
    "no-unused-vars": 0, // this gets handled by @typescript-eslint/no-unused-vars
    "@typescript-eslint/no-explicit-any": 0,
    "@typescript-eslint/prefer-as-const": 0
  }
}
