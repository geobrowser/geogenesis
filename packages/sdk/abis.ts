export * from './src/abis/index.js';
