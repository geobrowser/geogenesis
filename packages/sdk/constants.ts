export const INITIAL_COLLECTION_ITEM_INDEX_VALUE = 'a0';
