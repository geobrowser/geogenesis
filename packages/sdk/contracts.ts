// GOVERNANCE PLUGIN ADDRESSES
// from the plugin-repo-info.json in geo-browser-contracts after deploying
// Apr 25, 2024 â€“ L3 Testnet
export const SPACE_PLUGIN_REPO_ADDRESS =
  '0x3a93a9ACf3B57416A9aAd03A42C0b3C080BC808b'
export const GOVERNANCE_PLUGIN_REPO_ADDRESS =
  '0x8ff7cCad0f5edF47D28e49652D15AC826991b9ea'
export const PERSONAL_SPACE_ADMIN_PLUGIN_REPO_ADDRESS =
  '0x36b59d1b8E3323905d10b0E5e4A33E7cEe4E1c45'

export const DAO_FACTORY_ADDRESS = '0xb138AE700C352BB1aC75688e8ceCB98CDDaa7F09'
export const ENS_REGISTRY_ADDRESS = '0xD065A680075d0e27777bAa63CFFf17e1713a19Df'
