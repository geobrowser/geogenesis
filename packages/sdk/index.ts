export * from './src/types'

export { createGeoId, createTripleId } from './src/id';
export { getProcessGeoProposalArguments, getAcceptSubspaceArguments, getAcceptEditorArguments, getRemoveEditorArguments, getRemoveSubspaceArguments } from './src/encodings';
export { createCollection, createCollectionItem, reorderCollectionItem } from './src/collections';
export { SYSTEM_IDS } from './src/system-ids'
export { createImageEntityOps } from './src/create-image-entity'
