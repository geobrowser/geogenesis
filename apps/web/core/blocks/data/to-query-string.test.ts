import { SYSTEM_IDS } from '@geogenesis/sdk';
import { describe, expect, it } from 'vitest';

import { queryStringFromFilters } from './to-query-string';

describe('to-query-string', () => {
  it('Builds a graphql query from table block filters', () => {
    const stringFilter = queryStringFromFilters([
      {
        columnId: 'type',
        value: 'Value 1',
        valueType: 'TEXT',
      },
    ]);

    expect(stringFilter).toEqual(
      `triples: { some: { attributeId: { equalTo: "type" }, textValue: { equalToInsensitive: "Value 1"} } }`
    );

    const entityFilter = queryStringFromFilters([
      {
        columnId: 'type',
        value: 'id 1',
        valueType: 'RELATION',
      },
    ]);

    // Don't support relation filters in the query yet
    expect(entityFilter).toEqual(`and: []`);

    const nameFilter = queryStringFromFilters([
      {
        columnId: SYSTEM_IDS.NAME_ATTRIBUTE,
        value: 'id 1',
        valueType: 'TEXT',
      },
    ]);

    expect(nameFilter).toEqual(`name: { startsWithInsensitive: "id 1" }`);

    const andFilter = queryStringFromFilters([
      {
        columnId: SYSTEM_IDS.TYPES_ATTRIBUTE,
        value: 'Value 1',
        valueType: 'RELATION',
      },
      {
        columnId: SYSTEM_IDS.NAME_ATTRIBUTE,
        value: 'id 1',
        valueType: 'TEXT',
      },
    ]);

    expect(andFilter).toEqual(
      `and: [{ versionTypes: { some: { type: { entityId: {equalTo: "Value 1" } } } } }, { name: { startsWithInsensitive: "id 1" } }]`
    );

    const spaceFilter = queryStringFromFilters([
      {
        columnId: SYSTEM_IDS.SPACE_FILTER,
        valueType: 'TEXT',
        value: '0x0000000000000000000000000000000000000000',
      },
    ]);

    expect(spaceFilter).toEqual(
      `versionSpaces: {
        some: {
          spaceId: { equalTo: "0x0000000000000000000000000000000000000000" }
        }
      }`
    );

    const nullTypeIdFilter = queryStringFromFilters([]);

    expect(nullTypeIdFilter).toEqual('');
  });

  it('Builds a graphql query from table block filters for the postgraphile-based substreams API', () => {
    const stringFilter = queryStringFromFilters([
      {
        columnId: 'type',
        value: 'Value 1',
        valueType: 'TEXT',
      },
    ]);

    expect(stringFilter).toMatchSnapshot();

    const entityFilter = queryStringFromFilters([
      {
        columnId: 'type',
        value: 'id 1',
        valueType: 'RELATION',
      },
    ]);

    expect(entityFilter).toMatchSnapshot();

    const nameFilter = queryStringFromFilters([
      {
        columnId: SYSTEM_IDS.NAME_ATTRIBUTE,
        value: 'id 1',
        valueType: 'TEXT',
      },
    ]);

    expect(nameFilter).toMatchSnapshot();

    const andFilter = queryStringFromFilters([
      {
        columnId: 'type',
        value: 'Value 1',
        valueType: 'TEXT',
      },
      {
        columnId: 'type',
        value: 'id 1',
        valueType: 'RELATION',
      },
      {
        columnId: SYSTEM_IDS.NAME_ATTRIBUTE,
        value: 'id 1',
        valueType: 'TEXT',
      },
    ]);

    expect(andFilter).toMatchSnapshot();

    const spaceFilter = queryStringFromFilters([
      {
        columnId: SYSTEM_IDS.SPACE_FILTER,
        valueType: 'TEXT',
        value: '0x0000000000000000000000000000000000000000',
      },
    ]);

    expect(spaceFilter).toMatchSnapshot();

    const nullTypeIdFilter = queryStringFromFilters([]);

    expect(nullTypeIdFilter).toEqual('');
  });
});
