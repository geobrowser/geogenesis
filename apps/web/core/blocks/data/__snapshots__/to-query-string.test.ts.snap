// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`to-query-string > Builds a graphql query from table block filters for the postgraphile-based substreams API 1`] = `"triples: { some: { attributeId: { equalTo: "type" }, textValue: { equalToInsensitive: "Value 1"} } }"`;

exports[`to-query-string > Builds a graphql query from table block filters for the postgraphile-based substreams API 2`] = `
"relationsByFromVersionId: {
                some: {
                  typeOf: { id: { equalTo: "type" } }
                  toEntity: { id: { equalTo: "id 1" } }
                }
              }"
`;

exports[`to-query-string > Builds a graphql query from table block filters for the postgraphile-based substreams API 3`] = `"name: { startsWithInsensitive: "id 1" }"`;

exports[`to-query-string > Builds a graphql query from table block filters for the postgraphile-based substreams API 4`] = `
"and: [{ triples: { some: { attributeId: { equalTo: "type" }, textValue: { equalToInsensitive: "Value 1"} } } }, { relationsByFromVersionId: {
                some: {
                  typeOf: { id: { equalTo: "type" } }
                  toEntity: { id: { equalTo: "id 1" } }
                }
              } }, { name: { startsWithInsensitive: "id 1" } }]"
`;

exports[`to-query-string > Builds a graphql query from table block filters for the postgraphile-based substreams API 5`] = `"triples: { some: { attributeId: { equalTo: "JiFmyuFYeoiRSiY286m7A2" }, textValue: { equalToInsensitive: "0x0000000000000000000000000000000000000000"} } }"`;
